enhancements/olm-deletes-operands.md

```diff
@@ -0,0 +1,209 @@
+
+#### Operand cleanup API
+
+The CSV spec can have an optional field `spec.cleanup` which lets a user configure the cleanup behaviour before the CSV is deleted.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+...
+spec:
+  ...
+  cleanup:
+    enabled: false
+```
+
+With `spec.cleanup.enabled: true` OLM will first add the `olm/operand-cleanup` finalizer on the CSV. Now when the CSV is deleted by the user, OLM will delete all CRs for the operator's owned APIs, in the target namespaces of that operator. Once the CRs are successfully removed (after the operator clears its finalizers), OLM will clear the finalizer from the CSV to let the CSV and operator be uninstalled.
+
+With `spec.cleanup.enabled: false` OLM will remove the `olm/operand-cleanup` finalizer if it is already present on the CSV. This allows users to opt-out of cleanup and not have OLM delete any CRs.
+
+This is also the default behavior if `spec.cleanup` is unspecified. This ensures operator's that haven't opted-in are not subjected to the finalizer and have their CSV lifecycle remain unchanged.
+
+While toggling `spec.cleanup.enabled` does not result in a phase change, OLM could generate events to provide a history of when the cleanup behavior was toggled, and also events for when OLM begins and ends the cleanup process.
+
+#### Operand cleanup status
+
+To provide visibility on the status of operand cleanup, OLM can list out the CRs that are still pending deletion in the CSV status e.g `status.cleanup.pendingDeletion`.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+...
+status:
+  ...
+  cleanup:
+    pendingDeletion:
+    - resource: etcdclusters.etcd.database.coreos.com
+      kind: EtcdCluster
+      version: v1beta2
+      name: Foo
+      namespace: Bar
+    - kind: EtcdCluster
+      ...
+```
+
+This information can be useful when debugging a stalled operator uninstall process to see the CRs whose removal is blocking operand cleanup.
+
+**NOTE:** The number of CRs that OLM has to list in the status is unbounded and could potentially cause the CSV size to exceed the [max request size][max-request-size] configured in etcd. For this reason there would have to be an upper limit on the number of CRs displayed in `status.cleanup.pendingDeletion`.
+
+#### Handling CSV deletion due to replacement
+
+A CSV can also be [deleted by OLM][csv-deletion-on-upgrade] during it's lifecycle when OLM detects a newer CSV that replaces it as part of an upgrade. For an operator that has already opted-in to operand cleanup, this would result in OLM cleaning up all the CRs before removing the old CSV.
+
+To prevent the cleanup of CRs on an upgrade, OLM should ensure that a CSV which has `spec.cleanup.enabled: true` and is marked for replacement (`phase: replacing`), should first be cleared of the `olm/operand-cleanup` finalizer and opted-out of cleanup before it is marked for deletion or deleted.
+
```

Comment:
If a user enables/disables operand cleanup, should that be propagated to the upgraded CSV when an upgrade occurs?

Reply:
Hmm good question. I would think yes, since the user has already opted-in once and now expects that behavior.

Although I would defer to @kevinrizza to double check if there aren't cases where the upgraded operator might not support cleanup.
But for that we now have `spec.cleanup.supported` from the CSV author's side now to override and disable the cleanup behavior. See https://github.com/operator-framework/enhancements/pull/46#discussion_r525248931

Reply:
i would also say yes, the fields should be propagated to the upgraded CSV.

(this also might have been an argument for putting the field on the subscription instead, but i'm not proposing we revisit that decision)